src/sst/core/elemLoader.cc

```diff
@@ -135,92 +56,132 @@ splitPath(const std::string& searchPaths)
-    lt_handle = lt_dlopenadvise(libname.c_str(), loaderData->advise_handle);
-    if ( nullptr == lt_handle ) {
-        // The preopen module runs last and if the
-        // component was found earlier, but has a missing symbol or
-        // the like, we just get an amorphous "file not found" error,
-        // which is totally useless...
-        // Make darn sure we have enough space to hold the error message
-        std::vector<char> err_str(1e6);
-        sprintf(err_str.data(), "Opening element library %s failed: %s\n", elemlib.c_str(), lt_dlerror());
-        err_os << (const char*)err_str.data();
-        followError(libname, elemlib, searchPaths, err_os);
-    }
+    std::vector<std::string> paths = splitPath(searchPaths);
+
+    char* full_path     = new char[PATH_MAX];
+    bool  found_element = false;
+
+    for ( std::string& next_path : paths ) {
+        if ( verbose ) { printf("SST-DL: Searching: %s\n", next_path.c_str()); }
+
+        if ( next_path.at(next_path.size() - 1) == '/' ) {
+            sprintf(full_path, "%slib%s.so", next_path.c_str(), elemlib.c_str());
+        }
+        else {
+            sprintf(full_path, "%s/lib%s.so", next_path.c_str(), elemlib.c_str());
+        }
+
+        if ( verbose ) { printf("SST-DL: Attempting to load %s\n", full_path); }
+
+        // use a global bind policy read from environment, default to RTLD_LAZY
+        void* handle = dlopen(full_path, bindPolicy);
+
+#ifdef SST_COMPILE_MACOSX
+        // macOS will also allow files to use the dylib extension so this
+        // must also be checked. But only check this is the .so attempt
+        // failed first (because we may have had a successful load already)
+        // this implies ordering of .so before .dylib in priority.
+
+        if ( nullptr == handle ) {
+            if ( next_path.at(next_path.size() - 1) == '/' ) {
+                sprintf(full_path, "%slib%s.dylib", next_path.c_str(), elemlib.c_str());
+            }
+            else {
+                sprintf(full_path, "%s/lib%s.dylib", next_path.c_str(), elemlib.c_str());
+            }
+
+            if ( verbose ) { printf("SST-DL: Attempting to load %s\n", full_path); }
+
+            // use a global bind policy read from environment, default to RTLD_LAZY
+            handle = dlopen(full_path, bindPolicy);
+        }
+#endif
 
-    // loading a library can "wipe" previously loaded libraries depending
-    // on how weak symbol resolution works in dlopen
-    // rerun the loaders to make sure everything is still registered
-    for ( auto& libpair : ELI::LoadedLibraries::getLoaders() ) {
-        // loop all the elements in the element lib
-        for ( auto& elempair : libpair.second ) {
-            // loop all the loaders in the element
-            for ( auto* loader : elempair.second ) {
-                loader->load();
+        if ( nullptr == handle ) {
+            if ( verbose ) { printf("SST-DL: Loading failed, error: %s\n", dlerror()); }
```

Comment:
What happens here? Should we do something besides call `dlerror`?

Reply:
I believe we cannot fault/fatal because we can attempt to load modules in Python and fail. The code for explicit halts was removed some time ago.

Reply:
That was a feature Brandon was very keen on, but hasn't ever actually been used to my knowledge.  The expected behavior today if you import a module that doesn't exist (either because it doesn't find the shared library or the shared library doesn't have a python module) is that you get an exception.

Reply:
@feldergast so we should change this or leave it?

Reply:
I am suggesting that we leave this for now. I am not re-engineering all the handling of exceptions and crud in this PR. The point here is to keep PRs relatively small, focused - like this one, getting us out from `libtldl` so we can move forward with the CMake and no `libtool`. We should not need to re-engineer the core for every PR.

Reply:
I think the code as is will create the expected result, maybe plus or minus some console output. Further, it looks like there'll be enough information printed to help with debugging. Since no one really tries to import a library and then catches exceptions, I think this works for now.  If we start doing some of the exception handling, we may want to gather the text into a string to send with an exception, but we can revisit that later if it ever comes up.